### PR TITLE
exit group-by marking loop early

### PR DESCRIPTION
**Pull request by richardstartin**
For low cardinality one dimensional group by queries, it's possible to exit the group marking loop much earlier than the 10k default block size.

### CODE REVIEW

File: pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/groupby/DictionaryBasedGroupKeyGenerator.java [hunk 1]
```diff
@@ -270,6 +270,9 @@ private void processSingleValue(int numDocs, int[] dictIds, int[] outGroupIds) {
         if (!_flags[outGroupIds[i]]) {
           _numKeys++;
           _flags[outGroupIds[i]] = true;
+          if (_numKeys == _globalGroupIdUpperBound) {
```
Comment (siddharthteotia) [inline pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/groupby/DictionaryBasedGroupKeyGenerator.java:273]: Seems like the change was intended for conditional exit for low cardinality. I don't think that's what is happening here ?

Comment (richardstartin) [reply pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/groupby/DictionaryBasedGroupKeyGenerator.java:273]: This loop will only exit early when the cardinality is low.

Comment (Jackie-Jiang) [reply pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/groupby/DictionaryBasedGroupKeyGenerator.java:273]: We should also add one check before the for loop.

Out of the scope of this PR, in `processSingleValueGeneric()` and `processMultiValue()`, should we move `_flags[groupId] = true` into the if block?

**Review by Jackie-Jiang — APPROVED** [review]

File: pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/groupby/DictionaryBasedGroupKeyGenerator.java (hunk 1 above)
Comment (richardstartin) [reply pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/groupby/DictionaryBasedGroupKeyGenerator.java:273]: @Jackie-Jiang I'm not sure I follow - could you be more explicit please?

Comment (richardstartin) [reply pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/groupby/DictionaryBasedGroupKeyGenerator.java]: I think I got it, please see the latest commit

**Review by siddharthteotia — APPROVED** [review]

**Pull request merged**